pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
 // but instead use containerd to set it up, and xl to run it, so this does not need to do anything.
 // If/when we invert it, and have containerd launch the container with a microvm wrapper,
 // *then* this will need to work.
-func ctrStop(id string, force bool) error {
-	log.Infof("ctrStop %s %t\n", id, force)
+func ctrStop(containerID string, force bool) error {
+	log.Infof("ctrStop %s %t\n", containerID, force)
 	log.Infof("ctr stop done\n")
 	return nil
 }
 
 // ctrRm remove an existing container
-func ctrRm(id string) error {
-	log.Infof("ctrRm %s\n", id)
+func ctrRm(containerID string) error {
+	log.Infof("ctrRm %s\n", containerID)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return fmt.Errorf("Could not build containerd client. %v", err.Error())
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	container, err := loadContainer(containerID)
+	if err != nil {
+		return err
+	}
+	task, err := container.Task(ctrdCtx, cio.NewAttach(cio.WithStdio))
+	if err != nil {
+		log.Errorf("Could not fetch task of container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Could not fetch task of container: %v. %v", containerID, err.Error())
+	}
+	_, err = task.Delete(ctrdCtx)
+	if err != nil {
+		log.Errorf("Unable to delete task of container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Unable to delete task of container: %v. %v", containerID, err.Error())
+	}
+	err = container.Delete(ctrdCtx)
+	if err != nil {
+		log.Errorf("Unable to delete container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("Unable to delete container: %v. %v", containerID, err.Error())
+	}
 	log.Infof("ctrRm done\n")
 	return nil
 }
 
 // ctrCreate create a new container but do not start it
-func ctrCreate(id, imageHash string) error {
-	return nil
+func ctrCreate(containerID, imageName string, status *types.DomainStatus) (string, error) {
+	var (
+		ociOpts       []oci.SpecOpts
+		containerOpts []containerd.NewContainerOpts
+		ociSpec       oci.Spec
+		err           error
+	)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return "", fmt.Errorf("Exception while loading container. %v", err)
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	if len(status.EnvVariables) > 0 {
+		envVars := buildEnvVarsSlice(status.EnvVariables)
+		ociOpts = append(ociOpts, oci.WithEnv(envVars))
+	}
+
+	if len(ociOpts) > 0 {
+		containerOpts = append(containerOpts, containerd.WithSpec(&ociSpec, ociOpts...))
+	}
+	image, err := getImage(imageName)
```

Comment:
This seems wasteful. You already have a client and context in hand, why not just use it? It is a single line:

```go
	image, err := ctrdClient.GetImage(ctrdCtx, imageName)
```

Reply:
Corrected.